File: grida-domains/domain-rds/src/main/java/org/grida/entity/diaryimage/DiaryImageEntityRepository.java
package org.grida.entity.diaryimage;

import org.grida.domain.core.ImageDetail;
import org.grida.domain.diary.DiaryImage;
import org.grida.domain.diary.DiaryImageRepository;
import org.grida.exception.DomainRdsException;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Stream;

import static org.grida.exception.DomainRdsErrorCode.DIARY_IMAGE_NOT_FOUND;

@Repository
@Transactional(readOnly = true)
public class DiaryImageEntityRepository implements DiaryImageRepository {

    @PersistenceContext
    EntityManager entityManager;

    @Override
    @Transactional
    public long save(long diaryId, ImageDetail imageDetail, LocalDateTime lastActionAt) {
        DiaryImageEntity entity = DiaryImageMapper.toDiaryImageEntity(diaryId, imageDetail, lastActionAt);
        entityManager.persist(entity);

        return entity.getId();
    }

    @Override
    public DiaryImage findActivateImageByDiaryId(long diaryId) {
        return DiaryImageMapper.toDiaryImage(findActivateImageEntityByDiaryId(diaryId));
    }

    @Override
    public List<DiaryImage> findAllByDiaryId(long diaryId) {
        return findAllEntitiesStreamByDiaryId(diaryId)
                .map(DiaryImageMapper::toDiaryImage)
                .toList();
    }

    @Override
    @Transactional
    public void deactivateImage(long diaryId) {
        DiaryImageEntity entity = findActivateImageEntityByDiaryId(diaryId);
        entity.deactivate();
    }

    @Override
    @Transactional
    public void deleteAllByDiaryId(long diaryId) {
        findAllEntitiesStreamByDiaryId(diaryId)
                .forEach(entityManager::remove);
    }

    private DiaryImageEntity findActivateImageEntityByDiaryId(long diaryId) {
        return entityManager.createQuery(
                        "select di from DiaryImageEntity di " +
                                "where di.diaryId = : diaryId " +
                                "and di.isActivate = true",
                        DiaryImageEntity.class
                )
                .setParameter("diaryId", diaryId)
                .getResultStream()
                .findFirst()
                .orElseThrow(() -> new DomainRdsException(DIARY_IMAGE_NOT_FOUND));
    }

    private Stream<DiaryImageEntity> findAllEntitiesStreamByDiaryId(long diaryId) {
        return entityManager.createQuery(
                        "select di from DiaryImageEntity di " +
                                "where di.diaryId = : diaryId ",
                        DiaryImageEntity.class
                )
                .setParameter("diaryId", diaryId)
                .getResultStream();
    }
}

File: grida-database/database-rds/src/main/kotlin/org/grida/persistence/diaryimage/DiaryImageEntityRepository.kt
package org.grida.persistence.diaryimage

import org.grida.domain.diary.Diary
import org.grida.domain.diaryimage.DiaryImage
import org.grida.domain.diaryimage.DiaryImageRepository
import org.grida.domain.image.ImageStatus
import org.grida.domain.user.User
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
@Transactional(readOnly = true)
class DiaryImageEntityRepository(
    private val diaryImageJpaEntityRepository: DiaryImageJpaEntityRepository
) : DiaryImageRepository {

    @Transactional
    override fun save(
        diaryImage: DiaryImage,
        diary: Diary,
        user: User
    ): Long {
        val diaryEntity = diaryImage.toEntity(user, diary)
        diaryImageJpaEntityRepository.save(diaryEntity)
        return diaryEntity.id
    }

    override fun findById(id: Long): DiaryImage {
        val diaryImageEntity = diaryImageJpaEntityRepository.findByIdOrException(id)
        return diaryImageEntity.toDomain()
    }

    override fun findByDiaryIdAndStatus(diaryId: Long, status: ImageStatus): DiaryImage {
        val diaryImageEntity =
            diaryImageJpaEntityRepository.findByDiaryIdAndStatusOrException(diaryId, status)
        return diaryImageEntity.toDomain()
    }

    override fun countByDiaryId(diaryId: Long): Long {
        return diaryImageJpaEntityRepository.countByDiaryId(diaryId)
    }

    override fun existsByDiaryIdAndStatus(diaryId: Long, status: ImageStatus): Boolean {
        return diaryImageJpaEntityRepository.existsByDiaryIdAndStatus(diaryId, status)
    }

    @Transactional
    override fun updateStatus(
        diaryImageId: Long,
        status: ImageStatus
    ): Long {
        val diaryImageEntity = diaryImageJpaEntityRepository.findByIdOrException(diaryImageId)
        diaryImageEntity.updateStatue(status)
        return diaryImageEntity.id
    }
}
